Sources/NIOHTTP1/HTTPServerPipelineHandler.swift

```diff
@@ -92,53 +140,76 @@ public final class HTTPServerPipelineHandler: ChannelDuplexHandler, RemovableCha
             switch self {
+            case .preconditionFailed:
+                return .forceCloseConnection
             case .idle:
                 self = .requestAndResponseEndPending
+                return .none
             case .requestAndResponseEndPending, .responseEndPending, .requestEndPending,
                     .sentCloseOutputRequestEndPending, .sentCloseOutput:
-                preconditionFailure("received request head in state \(self)")
+                let message = "received request head in state \(self)"
+                self = .preconditionFailed
+                return .warnPreconditionViolated(message: message)
             }
         }
 
-        mutating func responseEndReceived() {
+        mutating func responseEndReceived() -> ConnectionStateAction {
             switch self {
+            case .preconditionFailed:
+                return .forceCloseConnection
             case .responseEndPending:
                 // Got the response we were waiting for.
                 self = .idle
+                return .none
             case .requestAndResponseEndPending:
                 // We got a response while still receiving a request, which we have to
                 // wait for.
                 self = .requestEndPending
+                return .none
             case .sentCloseOutput, .sentCloseOutputRequestEndPending:
                 // This is a user error: they have sent close(mode: .output), but are continuing to write.
                 // The write will fail, so we can allow it to pass.
-                ()
+                return .none
             case .requestEndPending, .idle:
-                preconditionFailure("Unexpectedly received a response in state \(self)")
+                let message = "Unexpectedly received a response in state \(self)"
+                self = .preconditionFailed
+                return .warnPreconditionViolated(message: message)
             }
         }
 
-        mutating func requestEndReceived() {
+        mutating func requestEndReceived() -> ConnectionStateAction {
             switch self {
+            case .preconditionFailed:
+                return .forceCloseConnection
             case .requestEndPending:
                 // Got the request end we were waiting for.
                 self = .idle
+                return .none
             case .requestAndResponseEndPending:
                 // We got a request and the response isn't done, wait for the
                 // response.
                 self = .responseEndPending
+                return .none
             case .sentCloseOutputRequestEndPending:
                 // Got the request end we were waiting for.
                 self = .sentCloseOutput
+                return .none
             case .responseEndPending, .idle, .sentCloseOutput:
-                preconditionFailure("Received second request")
+                let message = "Received second request"
```

Comment:
Also here in the `responseEndPending` case the peer violated the protocol. Should we really send an error down the pipeline first?

Reply:
The peer didn't violate the protocol because they can't. These messages are structured: they come from the parser, which cannot emit them in the wrong order. This error is a user error: it means they reordered the messages.